ambry-cloud/src/main/java/com/github/ambry/cloud/azure/AzureBlobDataAccessor.java

```diff
@@ -145,27 +146,33 @@ ProxyOptions getProxyOptions() {
    * @throws IOException for any error with supplied data stream.
    */
   public boolean uploadIfNotExists(BlobId blobId, long inputLength, CloudBlobMetadata cloudBlobMetadata,
-      InputStream blobInputStream) throws BlobStorageException, IOException {
+      InputStream blobInputStream) throws Exception {
     BlobRequestConditions blobRequestConditions = new BlobRequestConditions().setIfNoneMatch("*");
     azureMetrics.blobUploadRequestCount.inc();
     Timer.Context storageTimer = azureMetrics.blobUploadTime.time();
     try {
       Map<String, String> metadata = cloudBlobMetadata.toMap();
-      storageClient.uploadWithResponse(blobId, blobInputStream, inputLength, null, metadata, null, null,
-          blobRequestConditions, uploadTimeout);
+      storageClient.uploadAsyncWithResponse(blobId, blobInputStream, inputLength, null, metadata, null, null,
+          blobRequestConditions).get(uploadTimeout.toMillis(), TimeUnit.MILLISECONDS);
       logger.debug("Uploaded blob {} to ABS", blobId);
       azureMetrics.blobUploadSuccessCount.inc();
       return true;
-    } catch (UncheckedIOException e) {
-      // error processing input stream
-      throw e.getCause();
-    } catch (BlobStorageException e) {
-      if (e.getErrorCode() == BlobErrorCode.BLOB_ALREADY_EXISTS) {
-        logger.debug("Skipped upload of existing blob {}", blobId);
-        azureMetrics.blobUploadConflictCount.inc();
-        return false;
+    } catch (Exception e) {
+      Exception exception = Utils.extractFutureExceptionCause(e);
+      if (exception instanceof BlobStorageException) {
+        BlobStorageException blobStorageException = (BlobStorageException) exception;
+        if (blobStorageException.getErrorCode() == BlobErrorCode.BLOB_ALREADY_EXISTS) {
+          logger.debug("Skipped upload of existing blob {}", blobId);
+          azureMetrics.blobUploadConflictCount.inc();
+          return false;
+        } else {
+          throw blobStorageException;
+        }
+      } else if (exception instanceof UncheckedIOException) {
+        // error processing input stream
+        throw ((UncheckedIOException) exception).getCause();
       } else {
-        throw e;
+        throw exception;
```

Comment:
I think `Exception` is too generic to throw from here.

Reply:
It looks like we are currently checking for instance types of this exception when converting these exceptions to custom `CloudStorageException`.  I was wondering if we could probably handle more granular types there?

Reply:
Sorry, yeah may be it is too generic to catch and throw exceptions of type `Exception` class since it might also mask any checked exceptions (if they are added later by underneath layers). Modified it to throw subclass exceptions like `BlobStorageException`.